### PR TITLE
Implement ScalarValue::Map

### LINKED ISSUE #11128: Support Map as a ScalarValue
### Is your feature request related to a problem or challenge?

ScalarValue does not currently support Maps, see e.g. https://github.com/apache/datafusion/issues/8262#issuecomment-1852700799. 

https://github.com/apache/datafusion/issues/6485 is probably related as well, though I'm not 100% sure

I ran into this when looking to implement Substrait -> DF conversion for MapType

### Describe the solution you'd like

Implement ScalarValue::Map and all the related functionality

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by goldmedal**
## Which issue does this PR close?

Closes #11128 

## Rationale for this change
This PR only implements the Map in ScalarValue. However, we have no corresponding operations for it yet. I think we can implement a new scalar function called make_map() to create this scalar value. I plan to implement it in the next PR.

## Are these changes tested?
yes

## Are there any user-facing changes?
no

### CODE REVIEW

File: datafusion/common/src/scalar/mod.rs [hunk 1]
```diff
@@ -6199,6 +6339,46 @@ mod tests {
...
+
+        assert_eq!(map_value.to_string(), "[{joe:1},{blogs:2,foo:4},{},NULL]");
+
+        let ScalarValue::Map(arr) = map_value else {
+            panic!("Expected map");
+        };
+
+        //verify compared to arrow display
+        let batch = RecordBatch::try_from_iter(vec![("m", arr as _)]).unwrap();
+        let expected = [
```
Comment (alamb) [inline datafusion/common/src/scalar/mod.rs:6369]: 👍

**Review by alamb — APPROVED** [review]
Thank you @goldmedal  - I think this looks great to me

cc @Blizzara  what do you think?

File: datafusion/common/src/scalar/mod.rs [hunk 2]
```diff
@@ -3399,6 +3451,43 @@ impl fmt::Display for ScalarValue {
                         .join(",")
                 )?
             }
+            ScalarValue::Map(map_arr) => {
+                if map_arr.null_count() == map_arr.len() {
```
Comment (Blizzara) [inline datafusion/common/src/scalar/mod.rs:3455]: should this also have the assert like Struct? 
```
// ScalarValue Map should always have a single element
assert_eq!(map_arr.len(), 1);
```

File: datafusion/common/src/scalar/mod.rs [hunk 3]
```diff
@@ -3471,9 +3560,6 @@ impl fmt::Debug for ScalarValue {
             ScalarValue::List(_) => write!(f, "List({self})"),
             ScalarValue::LargeList(_) => write!(f, "LargeList({self})"),
             ScalarValue::Struct(struct_arr) => {
-                // ScalarValue Struct should always have a single element
-                assert_eq!(struct_arr.len(), 1);
-
```
Comment (Blizzara) [inline datafusion/common/src/scalar/mod.rs:3476]: was this removal intentional?

File: datafusion/common/src/scalar/mod.rs (hunk 2 above)
Comment (Blizzara) [reply datafusion/common/src/scalar/mod.rs:3455]: or if not, then maybe https://github.com/apache/datafusion/pull/11224/files#diff-49e275af8f09685c7bbc491db8ab3b9479960878f42ac558ec0e3e39570590bdR3583 shoulnd't have it either ? 😅

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs [hunk 4]
```diff
@@ -1246,6 +1246,30 @@ fn round_trip_scalar_values() {
...
+                "entries",
+                DataType::Struct(Fields::from(vec![
+                    Field::new("key", DataType::Int32, true),
+                    Field::new("value", DataType::Utf8, true),
+                ])),
+                false,
+            )),
+            true,
+        ))
+        .unwrap(),
```
Comment (Blizzara) [inline datafusion/proto/tests/cases/roundtrip_logical_plan.rs:1272]: these would both be NULL maps, right? is it possible to add a test for a map containing some values as well?

**Comment by Blizzara** [comment]
Thanks, looks good - I left couple notes but nothing major!

File: datafusion/common/src/scalar/mod.rs (hunk 2 above)
Comment (goldmedal) [reply datafusion/common/src/scalar/mod.rs:3455]: > should this also have the assert like Struct?
> 
> ```
> // ScalarValue Map should always have a single element
> assert_eq!(map_arr.len(), 1);
> ```


No, `MapArray` is `StructArray`. It could contain more than one element. 

> or if not, then maybe https://github.com/apache/datafusion/pull/11224/files#diff-49e275af8f09685c7bbc491db8ab3b9479960878f42ac558ec0e3e39570590bdR3583 shoulnd't have it either ? 😅

I forgot to remove this. Thanks for reminding me.

File: datafusion/common/src/scalar/mod.rs (hunk 3 above)
Comment (goldmedal) [reply datafusion/common/src/scalar/mod.rs:3476]: oops, I removed the wrong lines. I'll revert it. Thanks.

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs (hunk 4 above)
Comment (goldmedal) [reply datafusion/proto/tests/cases/roundtrip_logical_plan.rs:1272]: Sounds great. I'll add it. Thanks.

File: datafusion/common/src/scalar/mod.rs (hunk 2 above)
Comment (goldmedal) [reply datafusion/common/src/scalar/mod.rs:3455]: I add some tests to cover the Debug function in the latest commit.

**Review by Blizzara — APPROVED** [review]
LGTM!

**Comment by alamb** [comment]
Thanks @goldmedal  and @Blizzara  🚀

**Comment by goldmedal** [comment]
Thanks @alamb and @Blizzara !
